README.md
# Slippy Tile

[![Build Status](https://travis-ci.org/DenisCarriere/slippy-tile.svg?branch=master)](https://travis-ci.org/DenisCarriere/slippy-tile)
[![npm version](https://badge.fury.io/js/slippy-tile.svg)](https://badge.fury.io/js/slippy-tile)
[![Coverage Status](https://coveralls.io/repos/github/DenisCarriere/slippy-tile/badge.svg?branch=master)](https://coveralls.io/github/DenisCarriere/slippy-tile?branch=master)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/DenisCarriere/slippy-tile/master/LICENSE)
[![ES5](https://camo.githubusercontent.com/d341caa63123c99b79fda7f8efdc29b35f9f2e70/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f65732d352d627269676874677265656e2e737667)](http://kangax.github.io/compat-table/es5/)

[![Standard - JavaScript Style Guide](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

Slippy Tile helps parse a Tile scheme URL from a given Tile [x, y, zoom].

## Install

**npm**

```bash
$ npm install --save slippy-tile
```

**web**

```html
<script src="https://wzrd.in/standalone/slippy-tile@latest"></script>
```

## Quickstart

```javascript
var slippyTile = require('slippy-tile')
var tile = [10, 15, 8] // [x, y, zoom]
var scheme = 'https://{switch:a,b,c}.tile.openstreetmap.org/{zoom}/{x}/{y}.png'
var url = slippyTile(tile, scheme)
//= https://b.tile.openstreetmap.org/8/10/15.png
```

## Scheme

You can provide your own scheme by following the same syntax as JOSM.

**Options**

-   `{x}` or `{TileColumn}`: Tile Column
-   `{y}` or `{TileRow}`: Tile Row for Google Maps Compatible scheme
-   `{-y}`: Tile Row for TMS scheme
-   `{zoom}` or `{z}` or `{TileMatrix}`: Zoom Level
-   `{bbox}`: GeoJSON Bounding Box
-   `{bbox4326}`: GeoJSON Bounding Box using EPSG:4326 (WGS84)
-   `{bbox3857}`: GeoJSON Bounding Box using EPSG:3857 (World Mercator)
-   `{quadkey}` or `{q}`: Microsoft's Quadkey
-   `{switch:1,2,3}`: Selects a random sample
-   `{height}` default = 256
-   `{width}`: default = 256
-   `{size}`: default = 256,256
-   `{proj}`: default = EPSG:3857
-   `{format}`: default = image/png

## API

### slippyTile

Substitutes the given tile information [x, y, z] to the URL tile scheme.

**Parameters**

-   `tile` **Tile** Tile [x, y, z]
-   `url` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** URL Tile scheme or provider unique key
-   `options` **Object** Additional params
**Examples**

```javascript
slippyTile([10, 15, 8], 'https://{s}.tile.openstreetmap.org/{zoom}/{x}/{y}.png')
//='https://c.tile.openstreetmap.org/8/10/15.png'
```

Returns **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** parsed URL


CHANGELOG.md

# Changelog

## 4.0.0 - 2018-02-21

- Convert entire repo to Typescript
- Enforce strict Typescript configs & linting

## 3.0.1 - 2017-10-23

- Fix Typescript export default

## 3.0.0 - 2017-09-29

- Support ES Modules

## 2.3.0 - 2017-09-11

- Add new replacers
  - `{bbox4326}`: GeoJSON Bounding Box using EPSG:4326 (WGS84)
  - `{bbox3857}`: GeoJSON Bounding Box using EPSG:3857 (World Mercator)
  - `{size}`: default = 256,256

## 2.2.0 - 2017-09-08

- Dropped throw errors on uncomplete parsed URL
- Add extra optional params
- Remove extra devDependencies
- Support for NodeJS 4

## 2.1.0 - 2017-04-24

- Fix Bing Quadkey Regex parsing issue

## 2.0.0 - 2017-04-01

- Performance increase
- Drop providers in favor of `map-providers`
- **Breaking** Single function module `parse()` is now the only function for this library.

## 1.13.0 - 2017-03-27

- Add Toporama WMS as provider (English & French "not working")

## 1.12.0 - 2017-03-15

- Added `minZoom` & `maxZoom` to providers
- Improve rollup build
- Added minified build (11KB)

## 1.11.0 - 2017-03-08

- Convert module back to CommonJS
- Dropped Mapbox, DigitalGlobe & Strava (due to API keys being required)

## 1.10.0 - 2017-02-18

- Convert module to ES5
- Convert project to Rollup

## 1.9.0 - 2017-02-17

- Add case insenstive to replacements
- Replace `{bbox}` with meters instead of degrees

## 1.8.0 - 2017-02-14

- Convert library to ES2015

## 1.7.0 - 2017-01-19

- Support Webpack & Browserify
- Add HTML examples

## 1.6.0 - 2017-01-11

- Change OSM `{zoom}` to `{z}` as a safer URL
- Add individual parsing functions for WMS & WMTS
- Improve typescript handling for `sample`

## 1.5.0 - 2017-01-05

- Add schema support for WMS, WMTS, ArcGIS & Leaflet

## 1.4.2 - 2017-01-01

- Add providers as export

## 1.4.0 - 2016-12-04

- Added Circle CI tests
- Replace test cases with Jest
- Refactor providers to Typescript (removed .yml & .json)

## 1.3.2 - 2016-10-27

- Add WMTS 1.0.0 options `{TileColumn}`, `{TileRow}` & `{TileMatrix}`

## 1.3.0 - 2016-10-25

- Changed main function to `parse`
- Reduce library file size 550% (268KB to 48KB)

## 1.2.4 - 2016-10-21

- Add `{s}` to switch cases as default `['a', 'b', 'c']`
- Removed `js-yaml` dependency
- Removed all `request` related dependencies
- Removed `download` method
- Added DigitalGlobe as provider
- Converted `source.json` to `providers.yml`

## 1.1.0 - 2016-10-14

- Changed default function to `url`
- Added `download` feature
- Fallbacks, request-promise dependency for downloading Tile data Buffer
- Update docs

## 1.0.0 - 2016-10-13

First Stable release of `slippy-tile`.


index.ts
import { bboxToMeters, googleToBBox, googleToQuadkey, googleToTile } from "global-mercator";

export type Tile = [number, number, number];

/**
 * Substitutes the given tile information [x, y, z] to the URL tile scheme.
 *
 * @name slippyTile
 * @param {Tile} tile Tile [x, y, z]
 * @param {string} url URL Tile scheme or provider unique key
 * @param {Object} [options] Additional options
 * @param {string} [options.layer] Layer
 * @param {string} [options.version] Version
 * @param {string} [format="image/png"] Image Format
 * @returns {string} parsed URL
 * @example
 * slippyTile([10, 15, 8], "https://{s}.tile.openstreetmap.org/{zoom}/{x}/{y}.png")
 * //="https://c.tile.openstreetmap.org/8/10/15.png"
 */
function slippyTile(tile: Tile, url: string, options: {
  layer?: string,
  version?: string,
  format?: string,
  [key: string]: any,
} = {}) {
  const format = options.format || "image/png";
  const x = tile[0];
  const y = tile[1];
  const zoom = tile[2];
  url = wms(tile, url);
  url = wmts(url);
  url = parseSwitch(url);
  url = url.replace(/{(zoom|z|level)}/gi, String(zoom));
  url = url.replace(/{(x|col)}/gi, String(x));
  url = url.replace(/{(y|row)}/gi, String(y));
  url = url.replace(/{format}/gi, format);

  // Replace all additional options
  Object.keys(options).forEach((key) => {
    const pattern = RegExp("{" + key + "}", "ig");
    if (url.match(pattern)) {
      if (options[key] === undefined) { throw new Error("options." + key + " is required"); }
      url = url.replace(pattern, options[key]);
    }
  });
  if (url.match(/{-y}/)) { url = url.replace(/{-y}/gi, String(googleToTile(tile)[1])); }
  if (url.match(/{(quadkey|q)}/)) { url = url.replace(/{(quadkey|q)}/gi, googleToQuadkey(tile)); }
  return url;
}

/**
 * Parse WMS URL to friendly SlippyTile format
 *
 * @private
 * @param {Tile} tile Tile [x, y, z]
 * @param {string} url WMTS URL scheme
 * @returns {string} parsed URL
 * @example
 * wms([10, 15, 8], "/?layers=imagery&SRS={proj}&WIDTH={width}&HEIGHT={height}&BBOX={bbox}")
 * // => "/?layers=imagery&SRS=EPSG:3857&WIDTH=256&HEIGHT=256&BBOX=-165.9375,82.676285,-164.53125,82.853382"
 */
function wms(tile: Tile, url: string) {
  url = url.replace(/{height}/gi, "256");
  url = url.replace(/{width}/gi, "256");
  url = url.replace(/{size}/gi, "256,256");
  url = url.replace(/{(proj|srs|crs)}/gi, "EPSG:3857");
  const bbox = googleToBBox(tile);
  const bboxMeters = bboxToMeters(bbox);

  if (url.match(/EPSG:(3857|900913)/i) && url.match(/{bbox}/i)) { url = url.replace(/{bbox}/gi, bboxMeters.join(",")); }
  url = url.replace(/{bbox4326}/gi, bbox.join(","));
  url = url.replace(/{bbox3857}/gi, bboxMeters.join(","));
  url = url.replace(/{bbox}/gi, bbox.join(","));
  return url;
}

/**
 * Parse WMTS URL to friendly SlippyTile URL format
 *
 * @private
 * @param {string} url WMTS URL scheme
 * @returns {string} parsed URL
 * @example
 * wmts("https://<Tile Server>/WMTS/tile/1.0.0/Imagery/{Style}/{TileMatrixSet}/{TileMatrix}/{TileRow}/{TileCol}.jpg")
 * //="https://<Tile Server>/WMTS/tile/1.0.0/Imagery/default/GoogleMapsCompatible/{z}/{y}/{x}.jpg"
 */
function wmts(url: string) {
  if (url.match(/{TileCol|TileRow|TileMatrix|TileMatrixSet|Style}/gi)) {
    url = url.replace(/{TileCol}/gi, "{x}");
    url = url.replace(/{TileRow}/gi, "{y}");
    url = url.replace(/{TileMatrix}/gi, "{z}");
    url = url.replace(/{TileMatrixSet}/gi, "GoogleMapsCompatible");
    url = url.replace(/{Style}/gi, "default");
  }
  return url;
}

/**
 * Replaces {switch:a,b,c} with a random sample.
 *
 * @private
 * @param {string} url - URL Scheme
 * @returns {string} Parsed URL with switch replaced
 * @example
 * parseSwitch("http://tile-{switch:a,b,c}.openstreetmap.fr/hot/{zoom}/{x}/{y}.png")
 * //="http://tile-b.openstreetmap.fr/hot/{zoom}/{x}/{y}.png"
 */
function parseSwitch(url: string) {
  // Default simple switch
  if (url.match(/{s}/gi)) { return url.replace(/{s}/gi, String(sample(["a", "b", "c"]))); }

  // Custom switch
  const pattern = /{switch:([a-z,\d]*)}/i;
  const found = url.match(pattern);
  if (found) { return url.replace(pattern, String(sample(found[1].split(",")))); }
  return url;
}

/**
 * Sample an item from a given list
 *
 * @private
 * @name sample
 * @param {Array} array List of items
 * @returns {*} Single item from the list
 * @example
 * sample(["a", "b", "c"])
 * //="b"
 */
function sample(array: any[]) {
  if (array === null || array === undefined || array.length === 0) { return undefined; }
  return array[Math.floor(Math.random() * array.length)];
}

export default slippyTile;
